src/bench/pqs/spraylist.cpp

```diff
@@ -55,14 +55,38 @@ void
 spraylist::insert(const uint32_t &k,
                   const uint32_t &v)
 {
-    sl_add(m_q, k, READ_ADD_REM_ELASTIC_TX);
+    sl_add_val(m_q, k, v, TRANSACTIONAL);
+}
+
+void
+spraylist::insert(const size_t &k,
+                  const size_t &v)
+{
+    sl_add_val(m_q, k, v, TRANSACTIONAL);
 }
 
 bool
-spraylist::delete_min(uint32_t &v)
+spraylist::delete_min(size_t &k, size_t &v)
 {
-    const int ret = spray_delete_min(m_q, &v, d);
-    return (ret == 1);
+    unsigned long k_ret;
+    unsigned long v_ret;
+    int ret;
+    do {
+        ret = spray_delete_min_key(m_q, &k_ret, &v_ret, d);
```

Comment:
I will try to give answers to the first two of your gut reactions (please let me know if you don't agree with something):

1. I'm not sure if the old measurements in the random benchmark with the spraylist are incorrect given that you also measure the quality of the delete_min results. No retry loop in delete_min should give better performance but worse quality and the other way around. However, I think the chance that the results have relevance for a real application is larger with the retry loop as that is what one would like to have in a real application.
2. I think the reason you think it is weird to only have the retry loop for SparyList is that the retry loop should actually be inside SprayList's "own" delete_min function. I bet KLSM has some retry loop internally that instead could be handled by spurious failures, right? The spurious failures that the retry loop retries on are quite different from KLSM's spurious failures (I think) in how easily detectable they are.